### PR TITLE
Implement new vector types

### PR DESCRIPTION
**Pull request by gita-omr**
- remove VectorInt8 - VectorDouble from OMR::DataTypes
- allocate space for vector enums just before TR::NumAllTypes(former TR::NumTypes)
- vector enums are generated at runtime using createVectorType()
- value returned by createVectorType() can be used to dereference tables
  of size TR::NumAllTypes as before, but cannot be used to dereference
  tables of size TR::NumOMRTypes
- typeless vector opcodes such as vsub, vadd, etc. (have ILTypeProp::HasNoDataType property)
  can start using the new types
- codegen evaluators for those opcodes can use isVector(), getVectorElementType(), getVectorLength()
  on the node's DataType
- note that getVectorElementType() and getVectorLength() can only be called on vector types, otherwise there will be an assert
- to enable new vector length on specific platform, NumVectorLenghts must be defined corresponfingly in
  OMRDataTypes.hpp
- TRIL and JITBuilder will use 128-bit vectors for now

- ATTENTION: for safety reasons, renamed TR::NumTypes to TR::NumAllTypes. Please adjust your code correspondingly if it does not compile. In fact, any code that relies on specific values of TR::NumTypes, TR::NumOMRTypes, as well as vector opcodes being smaller than TR::NumOMRTypes will need to be adjusted. All that work is already done for openj9(https://github.com/eclipse-openj9/openj9/pull/14513).

### CODE REVIEW

**Comment by gita-omr** [comment]
@0xdaryl @jdmpapin @Leonardo2718  @mstoodle @vijaysun-omr could you please review the common part?
@Akira1Saitoh @r30shah @BradleyWood @IBMJimmyk could you please review aarch6, z, x, and p files correspondingly?

**Comment by gita-omr** [comment]
Updated commit message with a link to openj9 PR.

**Comment by gita-omr** [comment]
Looking at macOS failures...

**Comment by gita-omr** [comment]
Fixed a typo which somehow only showed up in macOS build.

**Comment by knn-k** [comment]
The changes in the aarch64 code look good to me.

File: compiler/aarch64/codegen/BinaryEvaluator.cpp
Comment (Akira1Saitoh) [inline compiler/aarch64/codegen/BinaryEvaluator.cpp:680]: The original code does not accept element types other than 8-bit integer, but there is no reason not to accept other integer types. Same for other logical bitwise operations (`vor`, `vxor` and `vnot`).

Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/BinaryEvaluator.cpp:680]: I agree. I will extend those evaluators once this PR gets merged.

File: compiler/x/codegen/OMRCodeGenerator.cpp
Comment (BradleyWood) [inline compiler/x/codegen/OMRCodeGenerator.cpp:991]: Shouldn't this accept ```VectorLength```

File: compiler/x/codegen/OMRTreeEvaluator.cpp
Comment (BradleyWood) [inline compiler/x/codegen/OMRTreeEvaluator.cpp:4069]: These tables only have entries for float & double, perhaps this can be simplified.

File: compiler/z/codegen/OMRCodeGenerator.cpp
Comment (BradleyWood) [inline compiler/z/codegen/OMRCodeGenerator.cpp:4725]: Use VectorLength?

File: compiler/z/codegen/SystemLinkageLinux.cpp
Comment (BradleyWood) [inline compiler/z/codegen/SystemLinkageLinux.cpp:596]: Remove the space

File: compiler/x/codegen/OMRCodeGenerator.cpp
Comment (gita-omr) [reply compiler/x/codegen/OMRCodeGenerator.cpp:991]: Yes, I agree. It would be better to accept VectorLength. I will change.

File: compiler/x/codegen/OMRTreeEvaluator.cpp
Comment (gita-omr) [reply compiler/x/codegen/OMRTreeEvaluator.cpp:4069]: Sure, but is it ok if we open another PR? I would like to minimize changes in this PR and make sure nothing is functionally broken. The main danger is that if I missed some table or some other place that needs to be changed...

File: compiler/x/codegen/OMRCodeGenerator.cpp
Comment (gita-omr) [reply compiler/x/codegen/OMRCodeGenerator.cpp]: Pushed commit with this change.

File: compiler/z/codegen/OMRCodeGenerator.cpp
Comment (gita-omr) [reply compiler/z/codegen/OMRCodeGenerator.cpp]: Pushed commit with this change.

**Comment by gita-omr** [comment]
Started 'win' test since there seemed to be environment failures.

File: compiler/x/codegen/OMRTreeEvaluator.cpp
Comment (BradleyWood) [reply compiler/x/codegen/OMRTreeEvaluator.cpp:4069]: Sure, I suppose that can wait

File: compiler/z/codegen/OMRLinkage.cpp
Comment (r30shah) [inline compiler/z/codegen/OMRLinkage.cpp:1998]: I am trying to understand why we have omitted FloatVector originally from having in arguments (Probably as short and extended format support for Vector instructions were added later in to the architecture). 
I am adding a `TODO` for myself here to double check and will confirm if that is the case, we might be able to remove check for element type here.

**Review by r30shah — CHANGES_REQUESTED** [review]
@gita-omr  I have briefly looked at the Z codegen changes and left comments. I will give it another look as well.

File: compiler/z/codegen/OMRLinkage.cpp
Comment (gita-omr) [reply compiler/z/codegen/OMRLinkage.cpp:1998]: Thanks! In some case above I already treat VectorFloat as all other vector types. Please let me know if should revert that.

File: compiler/z/codegen/SystemLinkagezOS.cpp
Comment (gita-omr) [reply compiler/z/codegen/SystemLinkagezOS.cpp:349]: Right, please let me know what I should do in this PR.

File: compiler/z/codegen/OMRTreeEvaluator.cpp
Comment (gita-omr) [reply compiler/z/codegen/OMRTreeEvaluator.cpp:16263]: Sure, I will improve it after you finish your review. Thanks a lot!

File: compiler/il/OMRDataTypes.hpp
Comment (0xdaryl) [inline compiler/il/OMRDataTypes.hpp:247]: "Scalar" is a more conventional term to use here than "NonVector"

File: compiler/il/OMRDataTypes.hpp
Comment (0xdaryl) [inline compiler/il/OMRDataTypes.hpp:420]: Please provide Doxygen comments for these APIs.

File: compiler/il/OMRDataTypes_inlines.hpp
Comment (0xdaryl) [inline compiler/il/OMRDataTypes_inlines.hpp:211]: Please use `static_cast` rather than the C-style cast.

File: compiler/il/OMRDataTypes_inlines.hpp
Comment (0xdaryl) [inline compiler/il/OMRDataTypes_inlines.hpp:221]: Please use `static_cast` rather than the C-style cast.

File: compiler/p/codegen/FPTreeEvaluator.cpp
Comment (0xdaryl) [inline compiler/p/codegen/FPTreeEvaluator.cpp:459]: May as well make this a fatal assert.

File: compiler/p/codegen/OMRCodeGenerator.cpp
Comment (0xdaryl) [inline compiler/p/codegen/OMRCodeGenerator.cpp]: Shouldn't the `length` parameter be a `TR::VectorLength` ?  That's how you defined it elsewhere.

File: compiler/z/codegen/OMRCodeGenerator.hpp
Comment (0xdaryl) [inline compiler/z/codegen/OMRCodeGenerator.hpp]: length should be `TR::VectorLength`

File: compiler/codegen/OMRCodeGenerator.hpp
Comment (0xdaryl) [inline compiler/codegen/OMRCodeGenerator.hpp]: length should be TR::VectorLength   ?

File: compiler/il/OMRDataTypes.hpp
Comment (0xdaryl) [reply compiler/il/OMRDataTypes.hpp:225]: oh, is it because of the need to set NumVectorLengths?

**Review by 0xdaryl — COMMENTED** [review]
I did a quick one-pass review just to get a sense of the scope of these changes.  I'll push out some initial comments now, but I do want to review this again in more detail.

File: compiler/il/OMRDataTypes.cpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes.cpp:99]: I am not sure who uses this method but that's what it did originally: returned integral type of the same element size as the input type.

File: compiler/il/OMRDataTypes.cpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes.cpp:250]: The names are cached in a static table so I think should be created only once per JIT execution.

File: compiler/il/OMRDataTypes.hpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes.hpp:225]: Yes, I put most commonly used lengths first.

File: compiler/il/OMRDataTypes.hpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes.hpp:247]: I wanted to bring attention to the existence of the new vector types. But we can rename to scalar if that's more preferable.

File: compiler/p/codegen/FPTreeEvaluator.cpp
Comment (gita-omr) [reply compiler/p/codegen/FPTreeEvaluator.cpp:459]: I would like to keep functional changes to the minimum in this PR. It was not FATAL assert before and I just mapped old interface into the equivalent new one.

File: compiler/p/codegen/OMRCodeGenerator.cpp
Comment (gita-omr) [reply compiler/p/codegen/OMRCodeGenerator.cpp]: That was already changed.

File: compiler/z/codegen/OMRCodeGenerator.hpp
Comment (gita-omr) [reply compiler/z/codegen/OMRCodeGenerator.hpp]: Old code. Fixed.

File: compiler/codegen/OMRCodeGenerator.hpp
Comment (gita-omr) [reply compiler/codegen/OMRCodeGenerator.hpp]: Already fixed.

File: compiler/z/codegen/OMRLinkage.cpp
Comment (gita-omr) [reply compiler/z/codegen/OMRLinkage.cpp:1998]: @r30shah just a reminder to let me know  how to treat VectorFloat in Z linkages.

File: compiler/il/OMRDataTypes.hpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes.hpp:420]: I thought names are pretty descriptive and the doxygen comments will make the code even a bit less readable. When all these methods are closely grouped together I think one can get a good idea on how to create and manipulate vector types. But, again, I can easily add those of course.

File: compiler/il/OMRDataTypes.cpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes.cpp]: fixed

File: compiler/il/OMRDataTypes_inlines.hpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes_inlines.hpp]: fixed

File: compiler/il/OMRDataTypes_inlines.hpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes_inlines.hpp]: fixed

**Comment by gita-omr** [comment]
Addressed all comments and squashed. Please resolve or respond to my questions.

File: compiler/il/OMRDataTypes.cpp
Comment (0xdaryl) [reply compiler/il/OMRDataTypes.cpp:99]: If all vector element types are equally valid then this seems wrong to me.  At the very least it should be understood why we can't simply return "self()->getDataType()" like the other element types rather than accepting that what's there now is correct.

File: compiler/il/OMRDataTypes.hpp
Comment (0xdaryl) [reply compiler/il/OMRDataTypes.hpp:225]: From a code readability POV, my opinion is to order them by increasing length.

File: compiler/il/OMRDataTypes.hpp
Comment (0xdaryl) [reply compiler/il/OMRDataTypes.hpp:247]: "Scalar" is the more preferable term to me.  Others may have different opinions.

File: compiler/il/OMRDataTypes.hpp
Comment (0xdaryl) [reply compiler/il/OMRDataTypes.hpp:420]: The established precedent for new or changed APIs in the OMR compiler has been to create Doxygen comments.

**Comment by gita-omr** [comment]
Rebased before addressing remaining comments.

File: compiler/z/codegen/SystemLinkagezOS.cpp
Comment (gita-omr) [reply compiler/z/codegen/SystemLinkagezOS.cpp:349]: Will preserve current VectorFloat behavior with next push.

File: compiler/z/codegen/OMRLinkage.cpp
Comment (gita-omr) [reply compiler/z/codegen/OMRLinkage.cpp:1998]: Please NVM. I will preserve current behavior.

File: compiler/z/codegen/OMRTreeEvaluator.cpp
Comment (gita-omr) [reply compiler/z/codegen/OMRTreeEvaluator.cpp:16263]: Will be fixed in next push.

File: compiler/z/codegen/OMRTreeEvaluator.cpp
Comment (gita-omr) [reply compiler/z/codegen/OMRTreeEvaluator.cpp:15653]: Will do.

File: compiler/z/codegen/OMRLinkage.cpp
Comment (gita-omr) [reply compiler/z/codegen/OMRLinkage.cpp:1998]: Will preserve current behavior.

File: compiler/il/OMRDataTypes.cpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes.cpp]: Right. Changed to jitPersistentAlloc and solved potential race condition (in the next push)

File: compiler/il/OMRDataTypes.hpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes.hpp:247]: Sure. Will make the change soon.

**Comment by knn-k** [comment]
Many files need updating the copyright year.
https://ci.eclipse.org/omr/job/CopyrightCheck/69/consoleText

**Comment by gita-omr** [comment]
I think I addressed all the comments but would like to start a personal build now.

**Comment by gita-omr** [comment]
Added isOMRDataType().

**Comment by gita-omr** [comment]
Fixed typo in the previous push.

**Comment by gita-omr** [comment]
While this PR was under review, some new code has been developed that uses old style vector types. Fixing it...

**Comment by gita-omr** [comment]
Resolved conflicts

**Comment by gita-omr** [comment]
Just a reminder that this PR will break most of the OMR consumers. Corresponding changes in openj9 are here: https://github.com/eclipse-openj9/openj9/pull/14513

**Comment by gita-omr** [comment]
Reviewed the code once again on the subject of any tables that are indexed by TR::DataTypes and therefore might need to be changed.  Applied some minor fixes.

**Comment by gita-omr** [comment]
Fixed white space.

**Comment by gita-omr** [comment]
Jenkins build all

**Comment by gita-omr** [comment]
Fixed windows build.

**Comment by gita-omr** [comment]
Jenkins build all

**Comment by gita-omr** [comment]
Windows build could not find` snprinf` so replaced it with `TR::snprintfNoTrunc`.

**Comment by gita-omr** [comment]
Jenkins build all

**Comment by gita-omr** [comment]
The PR is ready for a final review. Please take a look. Two questions still remain:

1) We decided to rename `TR::NumType`s into `TR::NumScalarTypes`. But those include `Aggregate` type. Should we rename it back to `TR::NumNonVectorType`?
2) Should we rename `TR::NumOMRTypes` into `TR::NumOMRNonVectorTypes` since this is what it represents? And also I do think we need to rename it into something to bring attention to the fact that the number of OMR non-vector types has shrunk (old vector types have been removed).

File: compiler/il/OMRIL.cpp
Comment (vijaysun-omr) [inline compiler/il/OMRIL.cpp:566]: If a read barrier is really a "load with some extra semantics" should we allow a `vreadbarrier` (given that we have a `vload`) ?

File: compiler/il/OMRIL.cpp
Comment (vijaysun-omr) [inline compiler/il/OMRIL.cpp:592]: Similar question to what I asked for read barriers earlier.

File: compiler/il/OMRDataTypes.cpp
Comment (gita-omr) [reply compiler/il/OMRDataTypes.cpp:273]: I used local static variable with initializer. So names will be created only once.

File: compiler/il/OMRIL.cpp
Comment (vijaysun-omr) [inline compiler/il/OMRIL.cpp:619]: Similar question as for the direct read barrier

File: compiler/il/OMRIL.cpp
Comment (vijaysun-omr) [inline compiler/il/OMRIL.cpp:645]: Similar question as for the direct write barrier

File: compiler/il/OMRIL.cpp
Comment (gita-omr) [reply compiler/il/OMRIL.cpp:592]: Please see the answer above.

File: compiler/il/OMRIL.cpp
Comment (gita-omr) [reply compiler/il/OMRIL.cpp:645]: The same.

File: compiler/il/OMRIL.cpp
Comment (vijaysun-omr) [inline compiler/il/OMRIL.cpp:688]: Why do we have all these different "vector register load" opcodes when we only have a single equivalent "vector load" opcode ?

**Comment by gita-omr** [comment]
Since a lot of files have been changed, I wanted to mention that the single core change is in OMRDataTypes.hpp. The rest is just an adjustment to that change and all current behaviour is preserved.

File: compiler/il/OMRIL.cpp
Comment (vijaysun-omr) [reply compiler/il/OMRIL.cpp:566]: I see, thanks. So this may be worth changing in the future but I agree with preserving behavior in this PR.

File: compiler/il/OMRIL.cpp
Comment (gita-omr) [reply compiler/il/OMRIL.cpp:688]: Actually, "New Vector Opcodes" PR is coming soon. It introduces new infrastructure for vector opcodes. Then, we will gradually remove old opcodes and add new ones.

File: compiler/il/OMROpcodes.enum
Comment (vijaysun-omr) [inline compiler/il/OMROpcodes.enum:6109]: Are these vector opcodes that may be re-designed in the future but are being left as per the original behavior in this PR ? Just wondering about the specific datatype in use for these (as opposed to the different approach for say a `vadd` opcode).

**Review by vijaysun-omr — APPROVED** [review]

**Comment by gita-omr** [comment]
Pushed a fix for OMR::DataType::getName()

File: compiler/il/OMROpcodes.enum
Comment (gita-omr) [reply compiler/il/OMROpcodes.enum:6109]: OMR::DataType::Vector128Int32 is a temporary static variable which will not be used.

Comment (vijaysun-omr) [reply compiler/il/OMROpcodes.enum:6109]: Thanks, I imagined this must be the case but wanted to double check with you. I have approved the PR since that was the last question I had from the review.

**Comment by 0xdaryl** [comment]
Reviewers:  would those who requested changes or made comments please formally approve this PR if you are satisfied with @gita-omr's resolution of them?

**Comment by 0xdaryl** [comment]
Jenkins build all

**Comment by knn-k** [comment]
There is a `TR::VectorDouble` remaining in OMR::X86::TreeEvaluator::vdsqrtEvaluator().
See https://github.com/eclipse-openj9/openj9/pull/14513#issuecomment-1082627203

**Comment by gita-omr** [comment]
> There is a `TR::VectorDouble` remaining in OMR::X86::TreeEvaluator::vdsqrtEvaluator(). See [eclipse-openj9/openj9#14513 (comment)](https://github.com/eclipse-openj9/openj9/pull/14513#issuecomment-1082627203)
> 
> The code was added by PR #6402, which was merged last week.

Right. Let me rebase and hopefully it gets merged soon :)

**Comment by gita-omr** [comment]
Fixed new code that was added while this PR was under review.

**Comment by 0xdaryl** [comment]
Jenkins build all

**Comment by 0xdaryl** [comment]
Jenkins build all

**Comment by gita-omr** [comment]
Fixed another use of an old vector type that was merged recently.

**Comment by BradleyWood** [comment]
Jenkins build all

**Pull request merged**
